File: Source/Tome.Model/Export/RecordExporter.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright file="RecordExporter.cs" company="Tome">
//   Copyright (c) Tome. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Tome.Model.Export
{
    using System.IO;
    using System.Linq;
    using System.Text;

    using Tome.Model.Project;

    public class RecordExporter
    {
        #region Constants

        private const string FieldIdPlaceholder = "$FIELD_ID$";

        private const string FieldTypePlaceholder = "$FIELD_TYPE$";

        private const string FieldValuePlaceholder = "$FIELD_VALUE$";

        private const string RecordFieldsPlaceholder = "$RECORD_FIELDS$";

        private const string RecordIdPlaceholder = "$RECORD_ID$";

        private const string RecordsPlaceholder = "$RECORDS$";

        #endregion

        #region Public Methods and Operators

        public void Export(TomeProject project, RecordExportTemplate template, string fileName)
        {
            // Build record file string.
            var recordStringBuilder = new StringBuilder();

            foreach (var recordFile in project.RecordFiles)
            {
                for (var recordIndex = 0; recordIndex < recordFile.Records.Count; ++recordIndex)
                {
                    var record = recordFile.Records[recordIndex];

                    // Build field values string.
                    var fieldValuesStringBuilder = new StringBuilder();
                    var fieldValues = record.FieldValues.ToList();

                    for (var fieldIndex = 0; fieldIndex < fieldValues.Count; ++fieldIndex)
                    {
                        var field = fieldValues[fieldIndex];

                        // Apply field value template.
                        var fieldDefinition =
                            project.FieldDefinitionFiles.SelectMany(file => file.FieldDefinitions)
                                .FirstOrDefault(definition => Equals(definition.Id, field.Key));

                        var fieldValueString = template.FieldValueTemplate;
                        fieldValueString = fieldValueString.Replace(FieldIdPlaceholder, field.Key);
                        fieldValueString = fieldValueString.Replace(
                            FieldTypePlaceholder,
                            fieldDefinition.FieldType.ToString());
                        fieldValueString = fieldValueString.Replace(FieldValuePlaceholder, field.Value.ToString());

                        fieldValuesStringBuilder.Append(fieldValueString);

                        // Add delimiter, if necessary.
                        if (fieldIndex < fieldValues.Count - 1)
                        {
                            fieldValuesStringBuilder.Append(template.FieldValueDelimiter);
                        }
                    }

                    // Apply record template.
                    var recordString = template.RecordTemplate;
                    recordString = recordString.Replace(RecordIdPlaceholder, record.Id);
                    recordString = recordString.Replace(RecordFieldsPlaceholder, fieldValuesStringBuilder.ToString());

                    recordStringBuilder.Append(recordString);

                    // Add delimiter, if necessary.
                    if (recordIndex < recordFile.Records.Count - 1)
                    {
                        recordStringBuilder.Append(template.RecordDelimiter);
                    }
                }
            }

            // Apply record file template.
            var recordFileString = template.RecordFileTemplate;
            recordFileString = recordFileString.Replace(RecordsPlaceholder, recordStringBuilder.ToString());

            // Write record file.
            var recordFileInfo = new FileInfo(fileName);

            using (var streamWriter = recordFileInfo.CreateText())
            {
                streamWriter.Write(recordFileString);
            }
        }

        #endregion
    }
}

File: Source/Tome/Export/recordexporter.cpp
#include "recordexporter.h"

using namespace Tome;

#include <QStringBuilder>
#include <QTextStream>

#include "../Values/valueconverter.h"


const QString RecordExporter::PlaceholderComponents = "$RECORD_COMPONENTS$";
const QString RecordExporter::PlaceholderComponentName = "$COMPONENT_NAME$";
const QString RecordExporter::PlaceholderFieldId = "$FIELD_ID$";
const QString RecordExporter::PlaceholderFieldType = "$FIELD_TYPE$";
const QString RecordExporter::PlaceholderFieldValue = "$FIELD_VALUE$";
const QString RecordExporter::PlaceholderRecordFields = "$RECORD_FIELDS$";
const QString RecordExporter::PlaceholderRecordId = "$RECORD_ID$";
const QString RecordExporter::PlaceholderRecords = "$RECORDS$";


RecordExporter::RecordExporter()
{
}

void RecordExporter::exportRecords(QSharedPointer<QIODevice> device, QSharedPointer<Project> project, QSharedPointer<RecordExportTemplate> exportTemplate)
{
    ValueConverter valueConverter;

    // Build record file string.
    QString recordsString;

    for (QVector<QSharedPointer<RecordSet> >::iterator itRecordSets = project->recordSets.begin();
         itRecordSets != project->recordSets.end();
         ++itRecordSets)
    {
        QSharedPointer<RecordSet> recordSet = *itRecordSets;

        for (QVector<QSharedPointer<Record> >::iterator itRecords = recordSet->records.begin();
             itRecords != recordSet->records.end();
             ++itRecords)
        {
            QSharedPointer<Record> record = *itRecords;

            // Build field values string.
            QString fieldValuesString;

            for (QMap<QString, QString>::iterator itFields = record->fieldValues.begin();
                 itFields != record->fieldValues.end();
                 ++itFields)
            {
                QString fieldId = itFields.key();
                QString fieldValue = itFields.value();
                QSharedPointer<FieldDefinition> fieldDefinition = project->getFieldDefinition(fieldId);

                // Apply field value template.
                QString fieldValueString = exportTemplate->fieldValueTemplate;
                fieldValueString = fieldValueString.replace(PlaceholderFieldId, fieldId);
                fieldValueString = fieldValueString.replace(PlaceholderFieldType, valueConverter.FieldTypeToString(fieldDefinition->fieldType));
                fieldValueString = fieldValueString.replace(PlaceholderFieldValue, fieldValue);

                fieldValuesString.append(fieldValueString);

                // Add delimiter, if necessary.
                if (itFields != record->fieldValues.end() - 1)
                {
                    fieldValuesString.append(exportTemplate->fieldValueDelimiter);
                }
            }

            // Collect components.
            QStringList components;

            for (QMap<QString, QString>::iterator itFields = record->fieldValues.begin();
                 itFields != record->fieldValues.end();
                 ++itFields)
            {
                QString fieldId = itFields.key();
                QSharedPointer<FieldDefinition> fieldDefinition = project->getFieldDefinition(fieldId);

                if (!fieldDefinition->component.isEmpty() && !components.contains(fieldDefinition->component))
                {
                    components.append(fieldDefinition->component);
                }
            }

            // Build components string.
            QString componentsString;

            for (QStringList::iterator itComponents = components.begin();
                 itComponents != components.end();
                 ++itComponents)
            {
                QString& component = *itComponents;

                // Apply component template.
                QString componentString = exportTemplate->componentTemplate;
                componentString = componentString.replace(PlaceholderComponentName, component);

                componentsString.append(componentString);

                // Add delimiter, if necessary.
                if (itComponents != components.end() - 1)
                {
                    componentsString.append(exportTemplate->componentDelimiter);
                }
            }

            // Apply record template.
            QString recordString = exportTemplate->recordTemplate;
            recordString = recordString.replace(PlaceholderRecordId, record->id);
            recordString = recordString.replace(PlaceholderRecordFields, fieldValuesString);
            recordString = recordString.replace(PlaceholderComponents, componentsString);

            recordsString.append(recordString);

            if (itRecords != recordSet->records.end() - 1 || itRecordSets != project->recordSets.end() - 1)
            {
                recordsString.append(exportTemplate->recordDelimiter);
            }
        }
    }

    // Apply record file template.
    QString recordFileString = exportTemplate->recordFileTemplate;
    recordFileString = recordFileString.replace(PlaceholderRecords, recordsString);

    // Write record file.
    QTextStream textStream(device.data());
    textStream << recordFileString;
}
